spinalcordtoolbox/reports/qc2.py

```diff
@@ -0,0 +1,359 @@
+
+        # Create a json file for the new QC report entry
+        path_json = path_qc / '_json'
+        path_json.mkdir(parents=True, exist_ok=True)
+        path_result = path_json / f'qc_{timestamp}.json'
+        with path_result.open('w') as file_result:
+            json.dump({
+                'cwd': str(Path.cwd()),
+                'cmdline': cmdline,
+                'command': command,
+                'sct_version': __version__,
+                'dataset': dataset,
+                'subject': subject,
+                'contrast': contrast,
+                'fname_in': path_input.name,
+                'plane': plane,
+                'background_img': str(imgs_to_generate['path_background_img'].relative_to(path_qc)),
+                'overlay_img': str(imgs_to_generate['path_overlay_img'].relative_to(path_qc)),
+                'moddate': mod_date.strftime("%Y-%m-%d %H:%M:%S"),
+                'qc': '',
+            }, file_result, indent=1)
+
+        # Collect all existing QC report entries
+        json_data = []
+        for path in sorted(path_json.glob('*.json')):
+            with path.open() as file:
+                json_data.append(json.load(file))
+
+        # Insert the QC report entries into index.html
+        path_assets = importlib.resources.files(spinalcordtoolbox.reports) / 'assets'
+        template = string.Template((path_assets / 'index.html').read_text(encoding='utf-8'))
+        # Empty the HTML file before writing, to make sure there's no leftover junk at the end
+        file_index_html.truncate()
+        file_index_html.write(template.substitute(sct_json_data=json.dumps(json_data)))
+
+        # Copy any missing assets
+        path_qc.mkdir(parents=True, exist_ok=True)
+        copy_missing(path_assets / '_assets', path_qc)
+
+    logger.info('Successfully generated the QC results in %s', str(path_result))
+    display_open(file=str(path_index_html), message="To see the results in a browser")
+
+
+def copy_missing(resource: Traversable, destination: Path):
+    """
+    Make sure that a copy of `resource` exists at `destination`,
+    by creating any missing files and directories recursively.
+    """
+    path = destination / resource.name
+    if resource.is_dir():
+        path.mkdir(exist_ok=True)
+        for sub_resource in resource.iterdir():
+            copy_missing(sub_resource, path)
+    elif resource.is_file():
+        if not path.is_file():
+            path.write_bytes(resource.read_bytes())
+    else:
+        # Some weird kind of resource? Ignore it
+        pass
+
+
+def sct_register_multimodal(
+    fname_input: str,
+    fname_output: str,
+    fname_seg: str,
+    argv: Sequence[str],
+    path_qc: str,
+    dataset: Optional[str],
+    subject: Optional[str],
+):
+    """
+    Generate a QC report for sct_register_multimodal.
+    """
+    command = 'sct_register_multimodal'
+    cmdline = [command]
+    cmdline.extend(argv)
```

Comment:
Minor nitpick regarding naming conventions:

Do we still want to call each `qc2` function name by its corresponding CLI script (e.g. `sct_register_multimodal`)? Or, do we want to create more "general" QC functions, in line with https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/4166, and how we wish to remove the `-p`/`-process` flag from `sct_qc`? 

I guess it's a little tricky, because to some degree we need to specify the `command`/`cmdline` argument. But, still, I'm wondering if this is the "final" organization we want to settle on?

Reply:
I don't think `qc2.sct_register_multimodal` is the last word, no! And I do want to support these "more general" QC functions down the line. But, I'm wary about premature abstraction, because I think that was the cause of the previous inflexibility in this code.

So, my main goal with this PR is to get the `with create_qc_entry` structure in place, which I think is pretty solid. After that, I envision a sequence of PRs where we "port" each existing QC report to this new structure, while de-abstracting the code as much as possible. As we do this, I hope that we can properly re-abstract the code, but only once 2 or (ideally) 3 instances of a given pattern appear. This should make it clear what functions to write, what their boundaries are, what their parameters are, and what their parameters *aren't*.

(This method is a little experimental for me, [inspired by a blog post](https://caseymuratori.com/blog_0015), and I think it will be worthwhile, but I'd like to have your approval before doing it. And if you find that the experiment is failing later on, I hope you'll tell me so that we can drop it.)